---
title: "**Survival Analyses with R**"
author: "Matt Maxwell"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document: 
   code folding: hide
   hard_wrap: TRUE
   theme: lumen
   toc: TRUE 
   toc_float: TRUE
   fig_caption: yes

---



<style>
.tocify {
  font-size: 11pt;
}

p.caption {
  font-size: 1.00em;
  color: black
}

pre {
  overflow-x: scroll
}

pre code {
  white-space: pre;
}

body {
  font-size: 12pt;
}

</style>

## **Introduction**
In this tutorial, I demo how to perform survival (AKA time to event) analysis with R for both human and mouse cancer datasets treated with checkpoint blockade immunotherapy. The emphasis of the analysis is in the context of my doctoral thesis work which focuses upon immunotherapy responsiveness of human and mouse tumors which harbor a mutation in the ARID1A tumor suppressor gene compared to tumors that do not. Following the demonstration survival analysis with log-rank test investigating a single variable, I also demonstrate how to perform a multi-variate analysis using the Cox regression model.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)

```

## **Import data & load packages**
Here we load the only two required packages for these analysis, the [Survival](https://github.com/therneau/survival) package for constructing our fitted survival model and [Survminer](https://github.com/kassambara/survminer) for visualizing the survival data with customizable Kaplan Meier survival curves. To acquire clinical trial data from Memorial Sloan Kettering (MSK) asessing response to immunotherapies such as anti-PDL1, anti-PD1, or anti-CTLA4, I visited the study page on the publicly available cBioPortal for cancer genomics and downloaded the [MSK data](https://www.cbioportal.org/study/summary?id=tmb_mskcc_2018) for all patients in the study (n=1661) as well as just the patients whose tumors have an ARID1A mutation (n=190). 
```{r MSK-Immunotherapy-Data-Import, message=FALSE, warning=FALSE}
library(survival)
library(survminer)


#To set working directory on windows machine, replace backslashes with double backslashes for R to accept our file path
wd <- r"(C:\Users\mattm\OneDrive\Desktop\GitHub_projects\Survival_analyses_with_R)"

#set our working directory (where our files are stored and outputted)
setwd(wd)

#set this as our working directory for all code chunks, not just this one.
#IGNORE if you're not using R markdown
knitr::opts_chunk$set(root.dir = wd)

#Set preference for no warnings to be given when running code chunks in R markdown
#IGNORE if you're not using R markdown
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

#read patient data from TMB MSKCC 2018 Immunotherapy study into Rstudio
survival_df <- read.delim("tmb_mskcc_2018_clinical_data.txt", header=T, sep = "\t" )
#replace single periods or double periods in column names with a single underscore
colnames(survival_df) <- gsub("\\.|\\.\\.", "_", colnames(survival_df))

head(survival_df)
dim(survival_df) #1661
length(unique(survival_df$Patient.ID)) #1661

#read in ARID1A mutant patient data
msk_arid1a <- read.delim("ARID1A_mutant_tmb_mskcc_2018_clinical_data.txt",header=T, sep = "\t")
colnames(msk_arid1a) <- gsub("\\.|\\.\\.", "_", colnames(msk_arid1a))
head(msk_arid1a)
dim(msk_arid1a) #190

```



## **Data wrangling for Kaplan Meier**
Here, I'm adding a new column to the `survival_df` dataframe to indicate whether each patient is ARID1A "non-mutant" or "mutant" by using the patient IDs from the 'msk_arid1a_id' object I created for ARID1A mutant patients. After this step, we're ready to visualize the survival differences between the two groups.
```{r Data-wrangling}

#Pull Patient IDs for ARID1A mutant patients
msk_arid1a_id <- as.vector(msk_arid1a$Sample_ID)

#assign a new column called ARID1A and give "non-mutant" as default value. We'll add mutant values in the next step.
survival_df$ARID1A <- "non-mutant"

#assign ARID1A mutant samples with "mutant" value in ARID1A column using the patient ID to match these patients to the larger dataset
survival_df$ARID1A[survival_df$Sample_ID %in% msk_arid1a_id] <- "mutant"
table(survival_df$ARID1A)

#  Mutant Non-mutant
#. 190      1468
```


## **Kaplan Meier curves**
Here, I use the data that I cleaned in the previous code chunk to fit a survival model of the data using the survfit() function from the survival package and plot it using the base R plot() function. This Kaplan Meier curve seems to show a difference in survival based on ARID1A mutation, but we could make this plot that is a lot more informative!
```{r basic-KM-curve, fig.cap = "**Figure 1: Kaplan Meier Curve of ARID1A Mutant vs Non-Mutant Human Cancers Following Immunotherapy**", fig.height=7, fig.width=7}

#Fit survival model using survfit() function from survival package using time (v14) and censored (v15) and specify that values of 1 are events (death).

#We also specify that we want to pull our data from the sample dataset and parse the data by values in the ARID1A column.
fit <- survfit(Surv(Overall_Survival_Months_, as.numeric(substr(Overall_Survival_Status, 1, 1))) ~ ARID1A,
  data = survival_df
)
plot(fit)

```


## **Customizing Kaplan Meier curves**
To make our kaplan meier curve plot more informative, let's add a color code, label the axes, a table below the plot to show the number of patients across time, a p-value from log-rank test, and confidence intervals. This should make for a much more informative kaplan meier plot! 
```{r MSK-Kaplan-Meier-curve, echo=FALSE, message=FALSE, warning=FALSE, fig.cap = "**Figure 2: Customized Kaplan Meier Curve of ARID1A Mutant vs Non-Mutant Human Cancers Following Immunotherapy**", fig.height=7, fig.width=7}

msk_surv <- ggsurvplot(fit, data = survival_df, palette = c("red", "blue"), title = "MSK Immunotherapy Cohort",
                       legend.title="Tumor Genotype", legend.labs=c("ARID1A mutant", "ARID1A non-mutant"), 
                       pval = TRUE, risk.table = TRUE, xlab = "Months", conf.int = TRUE)
msk_surv
```



## **Kaplan Meier curves - mouse data**
Here, I'm using survival data from mouse experiments from my PhD work where we compared the survival of mice that were injected with B16F10 melanoma or MC38 colon tumors with or without CRISPR-Cas9 mediated ARID1A deficiency (sgArid1a vs sgScramble) that were treated with checkpoint blockade immunotherapy. As you will see in the kaplan meier curves we generate below, the trend of significantly longer survival following immunotherapy in sgArid1a tumors vs sgScramble tumors is seen in both of our experimental tumor models. 
```{r B16F10-Immunotherapy, fig.cap = "**Figure 3: ARID1A Mutant B16F10 Melanoma Tumor Model Recapitulates Improved Immunotherapy Response**"}
b16_surv <- read.delim("B16F10_ICB_survival.txt", header = TRUE)

b16_fit <- survfit(Surv(time, cens) ~ group, data = b16_surv)

B16F10_surv <-ggsurvplot(b16_fit, xlim=c(0, 60), break.x.by = 12, ylab="Probability of Survival", xlab="Days post Injection",
           pval= TRUE, ylim=c(0,1),
           legend.labs=c("sgScramble", "sgArid1a"), legend.title="Tumor Genotype",  
           palette=c("blue", "red"),
           title="B16F10 ICB",  risk.table.height=.20)
B16F10_surv
```


```{r MC38-Immunotherapy, fig.cap = "**Figure 4: ARID1A Mutant MC38 Colon Tumor Model Recapitulates Improved Immunotherapy Response**"}
mc38_surv <- read.delim("MC38_ICB_survival.txt", header = TRUE)

mc38_fit <- survfit(Surv(time, cens) ~ group, data = mc38_surv)

MC38_surv <-ggsurvplot(mc38_fit, xlim=c(0, 60), break.x.by = 12, ylab="Probability of Survival", xlab="Days post Injection",
                       pval= TRUE, ylim=c(0,1),
                       legend.labs=c("sgScramble", "sgArid1a"), legend.title="Tumor Genotype",  
                      palette=c("blue", "red"),
                       title="MC38 ICB",  risk.table.height=.20)
MC38_surv
```
These experimental results suggest that the relationship between ARID1A mutation and improved immunotherapy response is causative rather than correlative. This is an important finding given the prevalance of ARID1A mutations in human cancers and gives credance to the recently posed hypothesis coming out of immunotherapy clinical trials that ARID1A mutation could serve as a biomarker of improved immunotherapy response in some cancers. While our experimental findings agree with this hypothesis, further clinical investigation will be needed to determine which scenarios ARID1A mutation status would serve as a useful biomarker.  


## **Cox multi-variate analysis**
Kaplan meier log rank test analysis is great for assessing survival differences between one variable such as ARID1A mutation status, but what about if we want to look at the impact of multiple variables on survival? Luckily, the Cox multi-variate regression model can handle this task and is relatively easy to implement. As an example, let's investigate the impact of ARID1A mutation status and biological sex on survival. 
```{r cox-survival, fig.cap = "**Figure 4: Cox multivariate regression model for ARID1A mutation status & biological sex in immunotherapy survival analysis**"}
#READ ME
#Before generating a cox model, we must convert character values in our columns representing our variables of interest to factors. Only character values need to be converted to factors, numeric values do not need to be converted to factors.
#For example, the ARID1A column character values of "non-mutant" or "mutant" must be converted to factors. 
survival_df$ARID1A <- as.factor(survival_df$ARID1A)
#The same need to convert to factors is true for the Sex column variable since the values of "Male" and "Female" are character values
survival_df$Sex <- as.factor(survival_df$Sex)

#Check levels of factors of interest
levels(survival_df$ARID1A)
levels(survival_df$Sex)


#Reorder ARID1A Factor levels so non-mutant is the reference factor for km & forest plots
new_levels <- c("non-mutant", "mutant")  # Replace with your desired order of levels
survival_df$ARID1A <- factor(survival_df$ARID1A, levels = new_levels)


# Create a cox regression model for ARID1A mutation and biological sex
cox_sex <- coxph(Surv(Overall_Survival_Months_, as.numeric(substr(Overall_Survival_Status, 1, 1))) ~ Sex + ARID1A, data = survival_df)

# Create cox data frame for possible combinations of ARID1A status and Sex status
sex_arid1a_df <- expand.grid(
  ARID1A = levels(survival_df$ARID1A),
  Sex = levels(survival_df$Sex))

#Assign arbitrary rowname values to the co-variate groupings
rownames(sex_arid1a_df) <- letters[1:4]
sex_arid1a_df

#Compute survival curves using survfit() function from survminer R package.
cxsurv <- survfit(cox_sex, data = survival_df, newdata = sex_arid1a_df, conf.type = "none")
#Check out the structure of cxsurv object
str(cxsurv)

#Make dataframe with survival curve information
surv_cxmod0 <- surv_summary(cxsurv)
head(surv_cxmod0)

#Append sex_arid1a_df rowname values to the surv_cxmod0 df
surv_cxmod <- cbind(surv_cxmod0,
                    sex_arid1a_df[as.character(surv_cxmod0$strata), ])

#Plot the survival curves for ARID1A and Sex variables
multivariate_survival <- ggsurvplot_df(surv_cxmod, linetype = "Sex"
                , color = "ARID1A",
                palette = c("red", "blue"),
                legend.title = NULL, censor = FALSE, risk.table = TRUE,
              pval = TRUE,
              conf.int = TRUE,
              legend = "bottom",
              title = "Survival Curves by ARID1A mutation & Biological Sex",
              xlab = "Overall Survival (Months)",
              ylab = "Survival Probability",
) 
multivariate_survival + theme(plot.title = element_text(hjust = 0.5))


```

## **Hazard ratios**
The kaplan meier plot visualizing the cox model is useful to get a sense of the impacts of the variables being analyzed, but what if we want to assess their statistical significance on survival? Conveniently, the `cox_sex` cox model we created earlier outputted hazard ratios, confidence intervals, and p values for our variables of interest which we can use to calculate their statistical significance. Hazard ratios are probabilistic expressions of the hazard rates for one group vs another group experiencing a given event which is death in this case. A hazard ratio of greater than 1 indicates a higher risk of experiencing death which is usually in comparing a treatment group to a control group. Accordingly, a hazard ratio less than 1 indicates a lower risk of death for a treatment group compared to a control group. Here we'll find out the hazard ratios associated with ARID1A mutation status and biological sex as well as the p values associated with these variables. 
```{r hazard-ratios}
#Take a look at the summary of our cox model
summary(cox_sex)
# Create a summary table of the model coefficients and their standard errors
summary_table <- summary(cox_sex)

# Extract the hazard ratios and their confidence intervals
hazard_ratios <- exp(summary_table$coefficients[, 1])
lower_ci <- exp(summary_table$coefficients[, 1] - 1.96 * summary_table$coefficients[, 2])
upper_ci <- exp(summary_table$coefficients[, 1] + 1.96 * summary_table$coefficients[, 2])

```
Above we see the summary of the `cox_sex` cox model. The hazard ratios are represented as exp(coef), the confidence intervals as the lower & upper .95 values, and p values from multiple statistical tests for any significant variables. Here, just ARID1A mutation as biological sex was not a significant determinant of survival.

## **Forest plot**
We can get all the information about hazard ratios, confidence intervals, and p values from our cox model summary, but we can quickly convey all of this information using a forest plot from the [survminer R package](https://www.rdocumentation.org/packages/survminer/versions/0.4.9/topics/ggforest).
```{r forest-plot, fig.cap = "**Figure 5: Forest plot for ARID1A mutation & Biological sex**", fig.width= 9}

# Create a data frame with the variable names, hazard ratios, and confidence intervals
forest_data <- data.frame(
  Variable = row.names(summary_table$coefficients),
  HazardRatio = hazard_ratios,
  LowerCI = lower_ci,
  UpperCI = upper_ci
)

ggforest(cox_sex, data = survival_df, fontsize = 0.8)

```

I used biological sex as a co-variate with ARID1A mutation because it's available in this dataset. The cox model here suggesting that males have a slightly better immunotherapy response (although not significant) than females could be due a number of confounding factors including the types of sex organ cancer included in the MSK dataset and thus biasing the sex variable towards males. To be clear, I'm only using biological sex as an example for co-variate analysis and don't plan on investigating immunotherapy response biases between sexes at present nor am I making any statements about sex and immunotherapy response, although this is a growing area of study in immunology. 

